File: app/src/main/kotlin/com/foodshare/features/challenges/presentation/ChallengesViewModel.kt
package com.foodshare.features.challenges.presentation

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.foodshare.core.errors.ErrorBridge
import com.foodshare.features.challenges.domain.model.*
import com.foodshare.features.challenges.domain.repository.ChallengeRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.auth.auth
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * UI State for Challenges screen.
 */
data class ChallengesUiState(
    val publishedChallenges: List<Challenge> = emptyList(),
    val userChallenges: List<ChallengeWithStatus> = emptyList(),
    val selectedChallenge: ChallengeWithStatus? = null,
    val leaderboard: List<ChallengeLeaderboardEntry> = emptyList(),
    val isLoading: Boolean = false,
    val isLoadingLeaderboard: Boolean = false,
    val isJoining: Boolean = false,
    val error: String? = null,
    val selectedFilter: ChallengeFilter = ChallengeFilter.ALL,
    val viewMode: ChallengeViewMode = ChallengeViewMode.LIST,
    val joinedChallengesCount: Int = 0,
    val completedChallengesCount: Int = 0,
    // Like states for optimistic updates
    val likeStates: Map<Int, Boolean> = emptyMap(),
    val likeCounts: Map<Int, Int> = emptyMap()
) {
    val filteredChallenges: List<ChallengeWithStatus>
        get() = when (selectedFilter) {
            ChallengeFilter.ALL -> userChallenges
            ChallengeFilter.JOINED -> userChallenges.filter {
                it.status == ChallengeUserStatus.ACCEPTED || it.status == ChallengeUserStatus.COMPLETED
            }
            ChallengeFilter.COMPLETED -> userChallenges.filter {
                it.status == ChallengeUserStatus.COMPLETED
            }
        }

    val deckChallenges: List<ChallengeWithStatus>
        get() = userChallenges.filter { it.status == ChallengeUserStatus.NOT_JOINED }
}

/**
 * ViewModel for Challenges feature.
 *
 * SYNC: This mirrors Swift ChallengesViewModel
 */
@HiltViewModel
class ChallengesViewModel @Inject constructor(
    private val repository: ChallengeRepository,
    private val supabaseClient: SupabaseClient
) : ViewModel() {

    private val _uiState = MutableStateFlow(ChallengesUiState())
    val uiState: StateFlow<ChallengesUiState> = _uiState.asStateFlow()

    private val currentUserId: String?
        get() = supabaseClient.auth.currentUserOrNull()?.id

    init {
        loadChallenges()
    }

    fun loadChallenges() {
        val userId = currentUserId ?: return
        if (_uiState.value.isLoading) return

        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }

            repository.getUserChallengesWithCounts(userId)
                .onSuccess { result ->
                    _uiState.update { state ->
                        // Initialize like states from challenges
                        val likeCounts = result.challenges.associate {
                            it.challenge.id to it.challenge.likesCount
                        }

                        state.copy(
                            userChallenges = result.challenges,
                            joinedChallengesCount = result.joinedCount,
                            completedChallengesCount = result.completedCount,
                            likeCounts = likeCounts,
                            isLoading = false
                        )
                    }

                    // Check like states for all challenges
                    result.challenges.forEach { challengeWithStatus ->
                        checkLikeStatus(challengeWithStatus.challenge.id)
                    }
                }
                .onFailure { error ->
                    _uiState.update {
                        it.copy(isLoading = false, error = ErrorBridge.mapChallengeError(error))
                    }
                }
        }
    }

    fun selectChallenge(challenge: Challenge) {
        val userId = currentUserId ?: return

        viewModelScope.launch {
            repository.getChallenge(challenge.id, userId)
                .onSuccess { challengeWithStatus ->
                    _uiState.update { it.copy(selectedChallenge = challengeWithStatus) }
                    loadLeaderboard(challenge.id)
                }
        }
    }

    fun clearSelectedChallenge() {
        _uiState.update { it.copy(selectedChallenge = null, leaderboard = emptyList()) }
    }

    fun acceptChallenge(challengeId: Int) {
        val userId = currentUserId ?: return
        if (_uiState.value.isJoining) return

        viewModelScope.launch {
            _uiState.update { it.copy(isJoining = true) }

            repository.acceptChallenge(challengeId, userId)
                .onSuccess {
                    loadChallenges() // Refresh to get updated counts
                }
                .onFailure { error ->
                    _uiState.update { it.copy(error = ErrorBridge.mapChallengeError(error)) }
                }

            _uiState.update { it.copy(isJoining = false) }
        }
    }

    fun completeChallenge(challengeId: Int) {
        val userId = currentUserId ?: return

        viewModelScope.launch {
            repository.completeChallenge(challengeId, userId)
                .onSuccess {
                    loadChallenges()
                }
                .onFailure { error ->
                    _uiState.update { it.copy(error = ErrorBridge.mapChallengeError(error)) }
                }
        }
    }

    fun rejectChallenge(challengeId: Int) {
        val userId = currentUserId ?: return

        viewModelScope.launch {
            repository.rejectChallenge(challengeId, userId)
                .onSuccess {
                    loadChallenges()
                }
                .onFailure { error ->
                    _uiState.update { it.copy(error = ErrorBridge.mapChallengeError(error)) }
                }
        }
    }

    fun loadLeaderboard(challengeId: Int) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoadingLeaderboard = true) }

            repository.getLeaderboard(challengeId, limit = 10)
                .onSuccess { entries ->
                    _uiState.update {
                        it.copy(leaderboard = entries, isLoadingLeaderboard = false)
                    }
                }
                .onFailure {
                    _uiState.update { it.copy(isLoadingLeaderboard = false) }
                }
        }
    }

    fun toggleLike(challengeId: Int) {
        val userId = currentUserId ?: return
        val currentState = _uiState.value

        // Optimistic update
        val currentlyLiked = currentState.likeStates[challengeId] ?: false
        val currentCount = currentState.likeCounts[challengeId] ?: 0
        val newLiked = !currentlyLiked
        val newCount = if (newLiked) currentCount + 1 else maxOf(0, currentCount - 1)

        _uiState.update { state ->
            state.copy(
                likeStates = state.likeStates + (challengeId to newLiked),
                likeCounts = state.likeCounts + (challengeId to newCount)
            )
        }

        viewModelScope.launch {
            repository.toggleLike(challengeId, userId)
                .onSuccess { (isLiked, likeCount) ->
                    // Server response - update with actual values
                    _uiState.update { state ->
                        state.copy(
                            likeStates = state.likeStates + (challengeId to isLiked),
                            likeCounts = state.likeCounts + (challengeId to likeCount)
                        )
                    }
                }
                .onFailure {
                    // Revert optimistic update
                    _uiState.update { state ->
                        state.copy(
                            likeStates = state.likeStates + (challengeId to currentlyLiked),
                            likeCounts = state.likeCounts + (challengeId to currentCount)
                        )
                    }
                }
        }
    }

    private fun checkLikeStatus(challengeId: Int) {
        val userId = currentUserId ?: return

        viewModelScope.launch {
            repository.hasLiked(challengeId, userId)
                .onSuccess { isLiked ->
                    _uiState.update { state ->
                        state.copy(likeStates = state.likeStates + (challengeId to isLiked))
                    }
                }
        }
    }

    fun setFilter(filter: ChallengeFilter) {
        _uiState.update { it.copy(selectedFilter = filter) }
    }

    fun setViewMode(mode: ChallengeViewMode) {
        _uiState.update { it.copy(viewMode = mode) }
    }

    fun clearError() {
        _uiState.update { it.copy(error = null) }
    }
}

File: Sources/FoodShare/Features/Challenges/Presentation/ViewModels/ChallengesViewModel.swift
//
//  ChallengesViewModel.swift
//  Foodshare
//
//  ViewModel for challenges feature
//

import Foundation
import Observation

@MainActor
@Observable
final class ChallengesViewModel {
    // MARK: - State

    var publishedChallenges: [Challenge] = []
    var userChallenges: [ChallengeWithStatus] = []
    var selectedChallenge: ChallengeWithStatus?
    var leaderboard: [ChallengeLeaderboardEntry] = []

    var isLoading = false
    var isLoadingLeaderboard = false
    var isJoining = false
    var error: AppError?
    var showError = false

    // MARK: - Like State

    var likeStates: [Int: Bool] = [:] // challengeId -> isLiked
    var likeCounts: [Int: Int] = [:] // challengeId -> count

    // MARK: - Filters

    var selectedFilter: ChallengeFilter = .all

    enum ChallengeFilter: String, CaseIterable {
        case all = "All"
        case joined = "Joined"
        case completed = "Completed"

        @MainActor
        func localizedDisplayName(using t: EnhancedTranslationService) -> String {
            switch self {
            case .all: t.t("challenges.filter.all")
            case .joined: t.t("challenges.filter.joined")
            case .completed: t.t("challenges.filter.completed")
            }
        }
    }

    // MARK: - Dependencies

    private let repository: ChallengeRepository
    private let userId: UUID?

    /// Whether the user is authenticated (has a userId)
    var isAuthenticated: Bool {
        userId != nil
    }

    // MARK: - Computed Properties

    var filteredChallenges: [Challenge] {
        switch selectedFilter {
        case .all:
            publishedChallenges
        case .joined:
            userChallenges.filter { $0.hasAccepted && !$0.hasCompleted }.map(\.challenge)
        case .completed:
            userChallenges.filter(\.hasCompleted).map(\.challenge)
        }
    }

    // Server-computed counts (avoid client-side .count(where:))
    private(set) var joinedChallengesCount = 0
    private(set) var completedChallengesCount = 0

    var errorMessage: String {
        error?.localizedDescription ?? "Failed to load challenges"
    }

    func localizedErrorMessage(using t: EnhancedTranslationService) -> String {
        error?.localizedDescription ?? t.t("challenges.load_failed")
    }

    // MARK: - Initialization

    init(repository: ChallengeRepository, userId: UUID? = nil) {
        self.repository = repository
        self.userId = userId
    }

    // MARK: - Actions

    func loadChallenges() async {
        guard !isLoading else { return }

        isLoading = true
        error = nil
        defer { isLoading = false }

        do {
            // Always load published challenges for all users
            publishedChallenges = try await repository.fetchPublishedChallenges()

            // Only load user-specific challenges if authenticated
            if let userId {
                // Use server-computed counts instead of client-side .count(where:)
                let result = try await repository.fetchUserChallengesWithCounts(userId: userId)
                userChallenges = result.challenges
                joinedChallengesCount = result.joinedCount
                completedChallengesCount = result.completedCount
            } else {
                userChallenges = []
                joinedChallengesCount = 0
                completedChallengesCount = 0
            }

            // Fetch translations for loaded challenges
            await fetchTranslationsForChallenges()

            await AppLogger.shared.info("Loaded \(publishedChallenges.count) challenges")
        } catch {
            self.error = .networkError(error.localizedDescription)
            showError = true
            await AppLogger.shared.error("Failed to load challenges", error: error)
        }
    }

    /// Fetch translations for current challenges from localization service
    private func fetchTranslationsForChallenges() async {
        let t = EnhancedTranslationService.shared

        // Skip if locale is English or no challenges
        guard t.currentLocale != "en", !publishedChallenges.isEmpty else { return }

        let challengeIds = publishedChallenges.map { Int64($0.id) }
        let translations = await t.fetchChallengeTranslations(challengeIds: challengeIds)

        guard !translations.isEmpty else { return }

        // Apply translations to published challenges
        for (index, challenge) in publishedChallenges.enumerated() {
            let challengeId = String(challenge.id)
            if let trans = translations[challengeId] {
                var updatedChallenge = challenge
                if let title = trans["title"] ?? nil {
                    updatedChallenge.titleTranslated = title
                }
                if let desc = trans["description"] ?? nil {
                    updatedChallenge.descriptionTranslated = desc
                }
                if updatedChallenge.titleTranslated != nil || updatedChallenge.descriptionTranslated != nil {
                    updatedChallenge.translationLocale = t.currentLocale
                    publishedChallenges[index] = updatedChallenge
                }
            }
        }

        // Also update challenges in userChallenges
        for (index, challengeWithStatus) in userChallenges.enumerated() {
            let challengeId = String(challengeWithStatus.challenge.id)
            if let trans = translations[challengeId] {
                var updatedChallenge = challengeWithStatus.challenge
                if let title = trans["title"] ?? nil {
                    updatedChallenge.titleTranslated = title
                }
                if let desc = trans["description"] ?? nil {
                    updatedChallenge.descriptionTranslated = desc
                }
                if updatedChallenge.titleTranslated != nil || updatedChallenge.descriptionTranslated != nil {
                    updatedChallenge.translationLocale = t.currentLocale
                    userChallenges[index] = ChallengeWithStatus(
                        challenge: updatedChallenge,
                        activity: challengeWithStatus.activity,
                        userId: challengeWithStatus.userId,
                    )
                }
            }
        }

        await AppLogger.shared.debug("Applied translations to \(translations.count) challenges")
    }

    func selectChallenge(_ challenge: Challenge) async {
        // For unauthenticated users, create a basic ChallengeWithStatus without activity
        guard let userId else {
            selectedChallenge = ChallengeWithStatus(challenge: challenge, activity: nil, userId: nil)
            await loadLeaderboard(for: challenge.id)
            return
        }

        do {
            selectedChallenge = try await repository.fetchChallenge(id: challenge.id, userId: userId)
            await loadLeaderboard(for: challenge.id)
        } catch {
            await AppLogger.shared.error("Failed to fetch challenge details", error: error)
        }
    }

    func acceptChallenge(_ challengeId: Int) async {
        guard let userId else {
            await AppLogger.shared.warning("Cannot accept challenge: user not authenticated")
            return
        }
        guard !isJoining else { return }

        isJoining = true
        defer { isJoining = false }

        do {
            _ = try await repository.acceptChallenge(challengeId: challengeId, userId: userId)
            HapticManager.success()

            // Refresh challenges
            await loadChallenges()

            // Update selected challenge if viewing
            if let challenge = selectedChallenge, challenge.challenge.id == challengeId {
                await selectChallenge(challenge.challenge)
            }

            await AppLogger.shared.info("Accepted challenge \(challengeId)")
        } catch {
            self.error = .networkError(error.localizedDescription)
            showError = true
            HapticManager.error()
            await AppLogger.shared.error("Failed to accept challenge", error: error)
        }
    }

    func completeChallenge(_ challengeId: Int) async {
        guard let userId else {
            await AppLogger.shared.warning("Cannot complete challenge: user not authenticated")
            return
        }

        do {
            _ = try await repository.completeChallenge(challengeId: challengeId, userId: userId)
            HapticManager.success()

            // Refresh challenges
            await loadChallenges()

            // Update selected challenge if viewing
            if let selected = selectedChallenge, selected.challenge.id == challengeId {
                await selectChallenge(selected.challenge)
            }

            await AppLogger.shared.info("Completed challenge \(challengeId)")
        } catch {
            self.error = .networkError(error.localizedDescription)
            showError = true
            HapticManager.error()
            await AppLogger.shared.error("Failed to complete challenge", error: error)
        }
    }

    func rejectChallenge(_ challengeId: Int) async {
        guard let userId else {
            await AppLogger.shared.warning("Cannot reject challenge: user not authenticated")
            return
        }

        do {
            _ = try await repository.rejectChallenge(challengeId: challengeId, userId: userId)
            HapticManager.light()

            // Refresh challenges
            await loadChallenges()

            await AppLogger.shared.info("Rejected challenge \(challengeId)")
        } catch {
            self.error = .networkError(error.localizedDescription)
            showError = true
            await AppLogger.shared.error("Failed to reject challenge", error: error)
        }
    }

    func loadLeaderboard(for challengeId: Int) async {
        guard !isLoadingLeaderboard else { return }

        isLoadingLeaderboard = true
        defer { isLoadingLeaderboard = false }

        do {
            leaderboard = try await repository.fetchLeaderboard(challengeId: challengeId, limit: 10)
        } catch {
            await AppLogger.shared.error("Failed to load leaderboard", error: error)
        }
    }

    func userStatus(for challenge: Challenge) -> ChallengeUserStatus {
        if let userChallenge = userChallenges.first(where: { $0.challenge.id == challenge.id }) {
            return userChallenge.status
        }
        return .notJoined
    }

    func dismissError() {
        error = nil
        showError = false
    }

    func refresh() async {
        await loadChallenges()
        HapticManager.light()
    }

    // MARK: - Like Actions
    // Note: Like toggling is now handled by ChallengeLikeButton component
    // using ChallengeEngagementService for consistency with Feed feature

    /// Check like status for a challenge using the centralized service
    func checkLikeStatus(for challengeId: Int) async {
        guard userId != nil else { return }

        do {
            let result = try await ChallengeEngagementService.shared.checkLiked(challengeId: challengeId)
            likeStates[challengeId] = result.isLiked
            likeCounts[challengeId] = result.likeCount
        } catch {
            // Silently fail - not critical
            await AppLogger.shared.debug("Failed to check like status for challenge \(challengeId)")
        }
    }

    func isLiked(challengeId: Int) -> Bool {
        likeStates[challengeId] ?? false
    }

    func likeCount(for challengeId: Int) -> Int {
        // Return cached count or fall back to challenge model count
        if let cached = likeCounts[challengeId] {
            return cached
        }
        return publishedChallenges.first(where: { $0.id == challengeId })?.challengeLikesCounter ?? 0
    }
}
